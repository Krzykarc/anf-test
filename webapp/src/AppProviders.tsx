import React from 'react';

interface AppProvidersProps {
}

export const AppProviders = (props: React.PropsWithChildren<AppProvidersProps>) => {
  const { children } = props

  // ðŸ”¥ w nastÄ™pnych moduÅ‚ach dojdÄ… providery
  return <>
    {children}
  </>
}
