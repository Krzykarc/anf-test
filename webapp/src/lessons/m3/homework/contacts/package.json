{
  "name": "anf-m3-homework-contacts",
  "private": true,
  "scripts": {
    "test": "jest --watch",
    "test:unit": "jest --ci"
  },
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "devDependencies": {
    "@babel/core": "^7.21.4",
    "@babel/preset-env": "^7.21.4",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.21.4",
    "@testing-library/jest-dom": "^5.16.5",
    "@types/enzyme": "^3.10.12",
    "@types/jest": "^29.5.0",
    "@types/node": "~16.11.68",
    "@types/react": "^17.0.58",
    "@types/react-dom": "^17.0.19",
    "@wojtekmaj/enzyme-adapter-react-17": "^0.8.0",
    "enzyme": "^3.11.0",
    "enzyme-to-json": "^3.6.2",
    "jest": "^29.5.0",
    "jest-environment-jsdom": "^29.5.0",
    "ts-node": "^10.9.1"
  }
}
