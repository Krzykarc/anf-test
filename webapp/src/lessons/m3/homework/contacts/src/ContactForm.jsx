/* eslint-disable */
import React from 'react';

export const ContactForm = (props) => {
 const defaultContact = {
    name: '',
    details: ''
  }

  const { contact = defaultContact } = props;

  const onChange = (e) => {
    let target = e.target;
    let name = target.name;
    let value = target.type === "checkbox" ? target.checked : target.value;

    props.onChange({
      ...contact,
      [name]: value
    });
  }

  const onSubmit = (e) => {
    e.preventDefault();
    props.onSubmit(props.contact);
  }

  return <>
    <form onSubmit={onSubmit}>
      <div className="form-group">
        <label className="d-block w-100">
          Name:
          <input
            className="form-control"
            name="name"
            value={props.contact.name}
            onChange={onChange}
          />
        </label>
      </div>
      <div className="form-group">
        <label className="d-block w-100">
          Contact details:
          <textarea
            className="form-control"
            name="details"
            value={props.contact.details}
            onChange={onChange}
          />
        </label>
      </div>
      <div className="form-group">
        <input
          type="button"
          id="cancel-contact"
          className="btn btn-default"
          value="Cancel"
          onClick={props.onCancel}
        />
        <input
          type="submit"
          id="save-contact"
          className="btn btn-primary ml-2"
          value="Save"
        />
      </div>
    </form>
  </>;
}
