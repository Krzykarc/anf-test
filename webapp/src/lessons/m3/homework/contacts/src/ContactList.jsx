/* eslint-disable */
import React from 'react';

export const ContactList = (props) => {
  const { contacts = [] } = props;

  return <>
    <div className="list-group">
      {contacts.map((contact) => (
        <button
          key={contact.id}
          onClick={(e) => props.onSelect(contact)}
          className={`list-group-item ${props.selected && (props.selected.id === contact.id) ? 'active' : ''}`}
        >
          {contact.name}
        </button>
      ))}
    </div>
  </>
}
