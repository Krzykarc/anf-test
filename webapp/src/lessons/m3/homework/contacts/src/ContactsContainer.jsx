/* eslint-disable */
import { useEffect, useState } from 'react';

import { ContactList } from './ContactList';
import { ContactForm } from './ContactForm';

export const ContactsContainer = (props) => {
  const { service } = props;

  const [ state, setState ] = useState({
    contacts: [],
    selected: null
  })

  useEffect(() => {
    service.getContacts().then((contacts) => {
      setState({
        contacts,
        selected: null,
      });
    });
  }, [])

  const newContact = () => {
    setState({
      contacts: [...state.contacts],
      selected: {
        name: '',
        details: ''
      }
    });
  }

  const onSelect = (contact) => {
    setState({
      contacts: [...state.contacts],
      selected: contact
    });
  }

  const onSubmit = (contact) => {
    service.saveContact(contact).then(()=>{
      service.getContacts().then((contacts) => {
        setState({
          contacts,
        });
      });
    });
  }

  const onCancel = () => {
    setState({
      contacts: [...state.contacts],
      selected: null
    });
  }

  return <>
    <div className="container">
      <div className="row">
        <div className="col-md-12">
          <h1>Contacts</h1>
        </div>
      </div>
      <div className="row">
        <div className="col-md-4">
          <ContactList
            contacts={state.contacts}
            selected={state.selected}
            onSelect={onSelect}
          />
        </div>
        <div className="col-md-4">
          {state.selected
          ? (
            <ContactForm contact={state.selected}
              onChange={onSelect}
              onSubmit={onSubmit}
              onCancel={onCancel}
            />
          ) : (
            <div>
              <button id="new-contact" onClick={newContact} className="btn btn-primary">New contact</button>
            </div>
          )}
        </div>
      </div>
    </div>
  </>;
}
