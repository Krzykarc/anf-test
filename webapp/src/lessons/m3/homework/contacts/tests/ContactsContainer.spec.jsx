import { fireEvent, render } from "@testing-library/react";

import flushPromises from 'flush-promises';
import { act } from "react-dom/test-utils";
import { mount } from 'enzyme';

import { ContactsContainer } from '../src/ContactsContainer';
import { ContactsService } from '../src/contacts.service';

describe('ContactsContainer Component', () => {
  it('a', () => {

  })
  function createMockService() {
    const data = [{
      id: '4567', name: 'candidate name', details: 'candidate details'
    }];

    const svc = new ContactsService(data);
    jest.spyOn(svc, 'getContacts');
    jest.spyOn(svc, 'saveContact');
    return svc;
  }

  it('should show candidates list when rendered', async () => {
    const mockService = createMockService();
    let findByText = null;
    await act(async () => {
      const wrapper = render(<ContactsContainer service={mockService} />);
      findByText = wrapper.findByText;
      await flushPromises();
    })

    expect(mockService.getContacts).toHaveBeenCalled();

    await flushPromises();

    await findByText('candidate name', {exact: false})
  });

  it('should call contactsService.saveContact() when form is submitted', async () => {
    const mockService = createMockService();
    let findByText = null;
    let findByLabelText = null;
    await act(async () => {
      const wrapper = render(<ContactsContainer service={mockService} />);
      findByText = wrapper.findByText;
      findByLabelText = wrapper.findByLabelText;
      await flushPromises();
    })

    expect(mockService.getContacts).toHaveBeenCalled();

    const contactItem = await findByText(mockService.contacts[0].name, {exact: false});
    fireEvent.click(contactItem);

    await act(async () => {
      const nameInput = await findByLabelText('Name', {exact: false, selector: 'input'});
      fireEvent.change(nameInput, {target: {value: 'Monika'}});
      await flushPromises();

      const detailsInput = await findByLabelText('Details', {exact: false, selector: 'textarea'});
      fireEvent.change(detailsInput, {target: {value: 'monia@gmail.com'}});
      await flushPromises();
  
      const saveButton = await findByText('Save');
      fireEvent.click(saveButton);
      await flushPromises();
    })

    expect(mockService.saveContact).toHaveBeenCalled();
  });

  it('should update contacts list when form is submitted', async () => {
    const mockService = createMockService();
    let findByText = null;
    let findByLabelText = null;

    await act(async () => {
      const wrapper = render(<ContactsContainer service={mockService} />);
      findByText = wrapper.findByText;
      findByLabelText = wrapper.findByLabelText;
      await flushPromises();
    })

    const newContactButton = await findByText('New contact', {exact: false});
    fireEvent.click(newContactButton);

    await act(async () => {
      const nameInput = await findByLabelText('Name', {exact: false, selector: 'input'});
      fireEvent.change(nameInput, {target: {value: 'Monika'}});
      await flushPromises();

      const detailsInput = await findByLabelText('Details', {exact: false, selector: 'textarea'});
      fireEvent.change(detailsInput, {target: {value: 'monia@gmail.com'}});
      await flushPromises();
  
      const saveButton = await findByText('Save');
      fireEvent.click(saveButton);
      await flushPromises();
    })

    await findByText('Monika', {exact: false});
  });
});
