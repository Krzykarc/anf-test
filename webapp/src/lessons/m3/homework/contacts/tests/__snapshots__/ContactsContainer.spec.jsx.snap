// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsContainer Component should match snapshot 1`] = `
<ContactsContainer
  service={
    ContactsService {
      "contacts": [
        {
          "details": "candidate test",
          "id": "4567",
          "name": "candidate test",
        },
      ],
      "getContacts": [MockFunction] {
        "calls": [
          [],
        ],
        "results": [
          {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "saveContact": [MockFunction],
    }
  }
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <h1>
          Contacts
        </h1>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-4"
      >
        <ContactList
          contacts={[]}
          onSelect={[Function]}
          selected={null}
        >
          <div
            className="list-group"
          />
        </ContactList>
      </div>
      <div
        className="col-md-4"
      >
        <div>
          <button
            className="btn btn-primary"
            id="new-contact"
            onClick={[Function]}
          >
            New contact
          </button>
        </div>
      </div>
    </div>
  </div>
</ContactsContainer>
`;
